src/sinks/nats.rs

```diff
@@ -0,0 +1,272 @@
+    SubjectTemplate { source: TemplateError },
+}
+
+/**
+ * Code dealing with the SinkConfig struct.
+ */
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct NatsSinkConfig {
+    encoding: EncodingConfigWithDefault<Encoding>,
+    #[serde(default = "default_name")]
+    name: String,
+    subject: String,
+    url: String,
+}
+
+fn default_name() -> String {
+    String::from("vector")
+}
+
+#[derive(Clone, Copy, Debug, Derivative, Deserialize, Serialize, Eq, PartialEq)]
+#[derivative(Default)]
+#[serde(rename_all = "snake_case")]
+pub enum Encoding {
+    #[derivative(Default)]
+    Text,
+    Json,
+}
+
+inventory::submit! {
+    SinkDescription::new::<NatsSinkConfig>("nats")
+}
+
+impl GenerateConfig for NatsSinkConfig {}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "nats")]
+impl SinkConfig for NatsSinkConfig {
+    async fn build(
+        &self,
+        cx: SinkContext,
+    ) -> crate::Result<(super::VectorSink, super::Healthcheck)> {
+        let sink = NatsSink::new(self.clone(), cx.acker())?;
+        let healthcheck = healthcheck(self.clone()).boxed();
+        Ok((super::VectorSink::Stream(Box::new(sink)), healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "nats"
+    }
+}
+
+impl NatsSinkConfig {
+    fn to_nats_options(&self) -> crate::Result<nats::Options> {
+        // Set reconnect_buffer_size on the nats client to 0 bytes so that the
+        // client doesn't buffer internally (to avoid message loss).
+
+        let options = nats::Options::new()
+            .with_name(&self.name)
+            .reconnect_buffer_size(0);
+
+        Ok(options)
+    }
+
+    async fn connect(&self) -> crate::Result<nats::asynk::Connection> {
+        self.to_nats_options()?
+            .connect_async(&self.url)
+            .map_err(|e| e.into())
+            .await
+    }
+}
+
+async fn healthcheck(config: NatsSinkConfig) -> crate::Result<()> {
+    config.connect().map_ok(|_| ()).await
+}
+
+/**
+ * Code dealing with the Sink struct.
+ */
+
+pub struct NatsSink {
+    encoding: EncodingConfig<Encoding>,
+    options: nats::Options,
+    subject: Template,
+    url: String,
+
+    acker: Acker,
+}
+
+impl NatsSink {
+    fn new(config: NatsSinkConfig, acker: Acker) -> crate::Result<Self> {
+        Ok(NatsSink {
+            acker,
+            options: config.to_nats_options()?,
+            subject: Template::try_from(config.subject).context(SubjectTemplate)?,
+            url: config.url,
+
+            // DEV: the following causes a move; needs to be last.
+            encoding: config.encoding.into(),
+        })
+    }
+}
+
+#[async_trait]
+impl StreamSink for NatsSink {
+    async fn run(&mut self, mut input: BoxStream<'_, Event>) -> Result<(), ()> {
+        // E0507: nats::Options does not implement Clone or Copy.
+        let nc = std::mem::replace(&mut self.options, nats::Options::new())
```

Comment:
This is just my inexperience with Rust shining through 😄 

`nats::Options` doesn't implement the `Clone` or `Copy` traits and I couldn't figure out how to satisfy the compiler in this instance..

Reply:
:smile: no worries; I'm still learning too. I think it's fine to just do `let nc = self.options.clone()` here instead. I think that'll satisfy that error. If not, could you provide the full error you are seeing?

Reply:
```
error[E0599]: no method named `clone` found for struct `nats::Options` in the current scope
   --> src/sinks/nats.rs:131:31
    |
131 |         let nc = self.options.clone()
    |                               ^^^^^ method not found in `nats::Options`
```

```
error[E0507]: cannot move out of `_self.options` which is behind a mutable reference
   --> src/sinks/nats.rs:131:19
    |
131 |         let nc = &self.options
    |                   ^^^^^^^^^^^^ move occurs because `_self.options` has type `nats::Options`, which does not implement the `Copy` trait
```

Reply:
```
error[E0507]: cannot move out of `_self.options` which is behind a mutable reference
   --> src/sinks/nats.rs:131:18
    |
131 |         let nc = self.options
    |                  ^^^^^^^^^^^^ move occurs because `_self.options` has type `nats::Options`, which does not implement the `Copy` trait
```

these were a few of the different techniques I tried

Reply:
@erhlee-bird

It would be easier if here:

```
pub struct NatsSink {
...
    options: nats::Options,
...
}
```

You kept the options in your own type. One where you can clone and do whatever you want. Then when you finally create the NATS connection you can do a 

`impl From<MyNatsOption> for nats::Options`

Reply:
@erhlee-bird

Here is what I meant: https://github.com/praveenperera/vector/commit/2f8fcd0aceb7d4e3417c0aea8f2b4eb965b815e4

Reply:
Thanks @praveenperera, that clears things up a lot.
I need to read about the From trait, but I get the gist of things